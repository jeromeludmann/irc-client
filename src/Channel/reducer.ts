import { combineReducers } from "redux";
import {
  CommandInputState,
  commandInputReducer,
} from "@app/CommandInput/reducers";
import { messageListReducer, MessageListState } from "@app/MessageList/reducer";

export interface ChannelState {
  messages: MessageListState;
  commandInput: CommandInputState;
}

export const channelReducer = combineReducers<ChannelState>({
  messages: messageListReducer,
  commandInput: commandInputReducer,
});
