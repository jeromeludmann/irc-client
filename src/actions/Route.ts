import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export type RoutedAction = Action & {
  route: Route;
};
