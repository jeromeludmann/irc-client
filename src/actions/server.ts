import { Action } from "redux";

export const CONNECT_SERVER = "SERVER/CONNECT_SERVER";
export const DISCONNECT_SERVER = "SERVER/DISCONNECT_SERVER";

export type ConnectServerAction = Action<typeof CONNECT_SERVER> & {
  payload: {
    host: string;
    port: number;
  };
};

export type DisconnectServerAction = Action<typeof DISCONNECT_SERVER>;

export const connectServer = (
  host: string,
  port: number = 6667,
): ConnectServerAction => ({
  type: CONNECT_SERVER,
  payload: { host, port },
});

export const disconnectServer = (): DisconnectServerAction => ({
  type: DISCONNECT_SERVER,
});
