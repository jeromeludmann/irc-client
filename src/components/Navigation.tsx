import React, { PureComponent, MouseEvent, CSSProperties } from "react";

interface Props {
  servers: {
    [s: string]: {
      windows: { [w: string]: { unread: boolean } };
    };
  };
  active: {
    server: string;
    window: string;
  };
  onWindowButtonClick: (
    { server, window }: { server: string; window: string },
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { active, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(server => (
          <ul key={server} style={{ listStyle: "none", padding: "0" }}>
            {Object.keys(servers[server].windows).map(window => {
              const isActive =
                server === active.server && window === active.window;
              const isUnread = servers[server].windows[window].unread;
              const style: CSSProperties = {
                outline: "none",
                border: "1px solid #ddd",
                width: "100px",
                color: isActive ? "black" : "#bbb",
                backgroundColor: isUnread ? "yellow" : "#fff",
              };

              return (
                <li key={window} style={{ marginBottom: "5px" }}>
                  <button
                    onClick={this.handleClick}
                    data-server={server}
                    data-window={window}
                    style={style}
                  >
                    {window}
                  </button>
                </li>
              );
            })}
          </ul>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const server = e.currentTarget.getAttribute("data-server");
    const window = e.currentTarget.getAttribute("data-window");

    if (server && window) {
      this.props.onWindowButtonClick({ server, window });
    }
  };
}
