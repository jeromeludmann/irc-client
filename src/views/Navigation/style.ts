import styled, { css } from 'styled-components'

export const List = styled.ul`
  margin-bottom: 10px;
  padding: 0;
  list-style: none;
`

export const ListItem = styled.li`
  margin: 5px 0;
  list-style: none;
`

interface ButtonProps {
  isActive: boolean
  hasActivity: boolean
}

export const Button = styled.button`
  outline: none;
  border: 1px solid #ddd;
  width: 100%;
  height: 30px;
  cursor: pointer;
  color: #bbb;
  background-color: white;
  font-weight: normal;
  &:hover {
    color: red;
  }

  ${(props: ButtonProps) =>
    props.isActive &&
    css`
      color: #333;
      font-weight: bold;
    `};

  ${(props: ButtonProps) =>
    props.hasActivity &&
    css`
      background-color: yellow;
    `};
`
