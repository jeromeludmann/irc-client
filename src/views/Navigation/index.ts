import { connect, MapStateToProps } from 'react-redux'
import { RouteState } from '@app/state/route/reducer'
import { Route } from '@app/utils/Route'
import { RootState } from '@app/state/root/reducer'
import { switchWindow } from '@app/actions/ui'
import Navigation from '@app/views/navigation/Navigation'
import { selectServers } from '@app/state/root/selectors'
import { selectRoute } from '@app/state/route/selectors'

interface StateProps {
  servers: {
    [serverKey: string]: {
      name: string
      buffers: { [bufferKey: string]: { activity: boolean } }
    }
  }
  window: RouteState
}

interface DispatchProps {
  onWindowButtonClick: (route: Route) => void
}

const mapStateToProps: MapStateToProps<StateProps, {}, RootState> = state => ({
  // TODO provide a better selector to Navigation
  servers: selectServers(state),
  window: selectRoute(state),
})

const mapDispatchToProps: DispatchProps = {
  onWindowButtonClick: switchWindow,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Navigation)
