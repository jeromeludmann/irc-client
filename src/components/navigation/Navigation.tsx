import React, { PureComponent, MouseEvent } from "react";
import { STATUS, RAW } from "@app/Route";
import { List, ListItem, Button } from "@app/components/navigation/style";

interface Props {
  servers: {
    [serverKey: string]: {
      name: string;
      channels: { [channelKey: string]: { activity: boolean } };
    };
  };
  window: {
    serverKey: string;
    channelKey: string;
  };
  onChannelButtonClick: (
    { serverKey, channelKey }: { serverKey: string; channelKey: string }
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { window, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(serverKey => (
          <List key={serverKey}>
            {Object.keys(servers[serverKey].channels).map(channelKey => (
              <ListItem key={channelKey}>
                <Button
                  onClick={this.handleClick}
                  data-server-key={serverKey}
                  data-channel-key={channelKey}
                  isActive={
                    serverKey === window.serverKey &&
                    channelKey === window.channelKey
                  }
                  hasActivity={servers[serverKey].channels[channelKey].activity}
                >
                  {channelKey === STATUS
                    ? servers[serverKey].name
                    : channelKey === RAW
                      ? serverKey
                      : channelKey}
                </Button>
              </ListItem>
            ))}
          </List>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const serverKey = e.currentTarget.getAttribute("data-server-key") as string;
    const channelKey = e.currentTarget.getAttribute(
      "data-channel-key"
    ) as string;

    this.props.onChannelButtonClick({ serverKey, channelKey });
  };
}
