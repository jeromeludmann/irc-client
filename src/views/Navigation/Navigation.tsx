import React, { PureComponent, MouseEvent } from 'react'
import { isStatus, isRaw } from '@app/utils/Route'
import { List, ListItem, Button } from '@app/views/navigation/style'

interface Props {
  servers: {
    [serverKey: string]: {
      name: string
      buffers: { [buffer: string]: { activity: boolean } }
    }
  }
  window: {
    serverKey: string
    bufferKey: string
  }
  onWindowButtonClick: (
    { serverKey, bufferKey }: { serverKey: string; bufferKey: string },
  ) => void
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { window, servers } = this.props

    return (
      <>
        {Object.keys(servers).map(serverKey => (
          <List key={serverKey}>
            {Object.keys(servers[serverKey].buffers).map(bufferKey => (
              <ListItem key={bufferKey}>
                <Button
                  onClick={this.handleClick}
                  data-server-key={serverKey}
                  data-buffer-key={bufferKey}
                  isActive={
                    serverKey === window.serverKey &&
                    bufferKey === window.bufferKey
                  }
                  hasActivity={servers[serverKey].buffers[bufferKey].activity}
                >
                  {isStatus(bufferKey)
                    ? servers[serverKey].name
                    : isRaw(bufferKey)
                      ? serverKey
                      : bufferKey}
                </Button>
              </ListItem>
            ))}
          </List>
        ))}
      </>
    )
  }

  private handleClick = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    this.props.onWindowButtonClick({
      serverKey: currentTarget.getAttribute('data-server-key')!,
      bufferKey: currentTarget.getAttribute('data-buffer-key')!,
    })
  }
}
