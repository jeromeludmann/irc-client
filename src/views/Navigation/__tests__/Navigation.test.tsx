import React from 'react'
import { shallow, mount } from 'enzyme'
import Navigation from '@app/views/navigation/Navigation'
import { BufferKey } from '@app/utils/Route'
import { bufferInitialState } from '@app/state/buffer/reducer'
import { serverInitialState } from '@app/state/server/reducer'

describe('Navigation component', () => {
  const props = {
    onWindowButtonClick: jest.fn(),
    servers: {
      serverKey: {
        name: 'irc.network',
        buffers: {
          ...serverInitialState.buffers,
          '#channel': { ...bufferInitialState, activity: true },
        },
      },
    },
    window: { serverKey: 'serverKey', bufferKey: BufferKey.STATUS },
  }

  it('should render correctly', () => {
    expect(shallow(<Navigation {...props} />)).toMatchSnapshot()
  })

  it('should handle button click event', () => {
    mount(<Navigation {...props} />)
      .find('button')
      .first()
      .simulate('click')
  })
})
