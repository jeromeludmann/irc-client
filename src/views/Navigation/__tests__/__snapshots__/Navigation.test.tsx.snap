// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navigation
    onWindowButtonClick={[MockFunction]}
    servers={
      Object {
        "serverKey": Object {
          "buffers": Object {
            "#channel": Object {
              "activity": true,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 0,
                  "values": Array [],
                },
                "value": "",
              },
              "messages": Array [],
            },
            "@raw": Object {
              "activity": false,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 0,
                  "values": Array [],
                },
                "value": "",
              },
              "messages": Array [],
            },
            "@status": Object {
              "activity": false,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 0,
                  "values": Array [],
                },
                "value": "",
              },
              "messages": Array [],
            },
          },
          "name": "irc.network",
        },
      }
    }
    window={
      Object {
        "bufferKey": "@status",
        "serverKey": "serverKey",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled.ul>
          <styled.li>
            <Styled(styled.button)
              data-buffer-key="@raw"
              data-server-key="serverKey"
              hasActivity={false}
              isActive={false}
              onClick={[Function]}
            >
              serverKey
            </Styled(styled.button)>
          </styled.li>
          <styled.li>
            <Styled(styled.button)
              data-buffer-key="@status"
              data-server-key="serverKey"
              hasActivity={false}
              isActive={true}
              onClick={[Function]}
            >
              irc.network
            </Styled(styled.button)>
          </styled.li>
          <styled.li>
            <Styled(styled.button)
              data-buffer-key="#channel"
              data-server-key="serverKey"
              hasActivity={true}
              isActive={false}
              onClick={[Function]}
            >
              #channel
            </Styled(styled.button)>
          </styled.li>
        </styled.ul>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "serverKey",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.li>
              <Styled(styled.button)
                data-buffer-key="@raw"
                data-server-key="serverKey"
                hasActivity={false}
                isActive={false}
                onClick={[Function]}
              >
                serverKey
              </Styled(styled.button)>
            </styled.li>,
            <styled.li>
              <Styled(styled.button)
                data-buffer-key="@status"
                data-server-key="serverKey"
                hasActivity={false}
                isActive={true}
                onClick={[Function]}
              >
                irc.network
              </Styled(styled.button)>
            </styled.li>,
            <styled.li>
              <Styled(styled.button)
                data-buffer-key="#channel"
                data-server-key="serverKey"
                hasActivity={true}
                isActive={false}
                onClick={[Function]}
              >
                #channel
              </Styled(styled.button)>
            </styled.li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "@raw",
            "nodeType": "class",
            "props": Object {
              "children": <Styled(styled.button)
                data-buffer-key="@raw"
                data-server-key="serverKey"
                hasActivity={false}
                isActive={false}
                onClick={[Function]}
              >
                serverKey
              </Styled(styled.button)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "serverKey",
                "data-buffer-key": "@raw",
                "data-server-key": "serverKey",
                "hasActivity": false,
                "isActive": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "serverKey",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "@status",
            "nodeType": "class",
            "props": Object {
              "children": <Styled(styled.button)
                data-buffer-key="@status"
                data-server-key="serverKey"
                hasActivity={false}
                isActive={true}
                onClick={[Function]}
              >
                irc.network
              </Styled(styled.button)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "irc.network",
                "data-buffer-key": "@status",
                "data-server-key": "serverKey",
                "hasActivity": false,
                "isActive": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "irc.network",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "#channel",
            "nodeType": "class",
            "props": Object {
              "children": <Styled(styled.button)
                data-buffer-key="#channel"
                data-server-key="serverKey"
                hasActivity={true}
                isActive={false}
                onClick={[Function]}
              >
                #channel
              </Styled(styled.button)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "#channel",
                "data-buffer-key": "#channel",
                "data-server-key": "serverKey",
                "hasActivity": true,
                "isActive": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "#channel",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled.ul>
            <styled.li>
              <Styled(styled.button)
                data-buffer-key="@raw"
                data-server-key="serverKey"
                hasActivity={false}
                isActive={false}
                onClick={[Function]}
              >
                serverKey
              </Styled(styled.button)>
            </styled.li>
            <styled.li>
              <Styled(styled.button)
                data-buffer-key="@status"
                data-server-key="serverKey"
                hasActivity={false}
                isActive={true}
                onClick={[Function]}
              >
                irc.network
              </Styled(styled.button)>
            </styled.li>
            <styled.li>
              <Styled(styled.button)
                data-buffer-key="#channel"
                data-server-key="serverKey"
                hasActivity={true}
                isActive={false}
                onClick={[Function]}
              >
                #channel
              </Styled(styled.button)>
            </styled.li>
          </styled.ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "serverKey",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.li>
                <Styled(styled.button)
                  data-buffer-key="@raw"
                  data-server-key="serverKey"
                  hasActivity={false}
                  isActive={false}
                  onClick={[Function]}
                >
                  serverKey
                </Styled(styled.button)>
              </styled.li>,
              <styled.li>
                <Styled(styled.button)
                  data-buffer-key="@status"
                  data-server-key="serverKey"
                  hasActivity={false}
                  isActive={true}
                  onClick={[Function]}
                >
                  irc.network
                </Styled(styled.button)>
              </styled.li>,
              <styled.li>
                <Styled(styled.button)
                  data-buffer-key="#channel"
                  data-server-key="serverKey"
                  hasActivity={true}
                  isActive={false}
                  onClick={[Function]}
                >
                  #channel
                </Styled(styled.button)>
              </styled.li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "@raw",
              "nodeType": "class",
              "props": Object {
                "children": <Styled(styled.button)
                  data-buffer-key="@raw"
                  data-server-key="serverKey"
                  hasActivity={false}
                  isActive={false}
                  onClick={[Function]}
                >
                  serverKey
                </Styled(styled.button)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "serverKey",
                  "data-buffer-key": "@raw",
                  "data-server-key": "serverKey",
                  "hasActivity": false,
                  "isActive": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "serverKey",
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "@status",
              "nodeType": "class",
              "props": Object {
                "children": <Styled(styled.button)
                  data-buffer-key="@status"
                  data-server-key="serverKey"
                  hasActivity={false}
                  isActive={true}
                  onClick={[Function]}
                >
                  irc.network
                </Styled(styled.button)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "irc.network",
                  "data-buffer-key": "@status",
                  "data-server-key": "serverKey",
                  "hasActivity": false,
                  "isActive": true,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "irc.network",
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "#channel",
              "nodeType": "class",
              "props": Object {
                "children": <Styled(styled.button)
                  data-buffer-key="#channel"
                  data-server-key="serverKey"
                  hasActivity={true}
                  isActive={false}
                  onClick={[Function]}
                >
                  #channel
                </Styled(styled.button)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "#channel",
                  "data-buffer-key": "#channel",
                  "data-server-key": "serverKey",
                  "hasActivity": true,
                  "isActive": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "#channel",
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
